docs/language/learn-ql/cpp/value-numbering-hash-cons.rst

```diff
@@ -0,0 +1,99 @@
+
+The value numbering API
+~~~~~~~~~~~~~~~~~~~~~~~
+The value numbering library exposes its interface primarily through the ``GVN`` class. Each instance of ``GVN`` represents a set of expressions that will always evaluate to the same value. To get an expression in the set represented by a particular ``GVN``, use the ``getAnExpr()`` member predicate.
+
+To get the ``GVN`` of an ``Expr``, use the ``globalValueNumber`` predicate.
+
+.. note::
+
+    While the ``GVN`` class has ``toString`` and ``getLocation`` methods, these are only provided as debugging aids. They give the ``toString`` and ``getLocation`` of an arbitrary ``Expr`` within the set.
+
+Example Queries
+~~~~~~~~~~~~~~~
+
+This query uses the ``GVN`` class to identify calls to ``strncpy`` where the size argument is derived from the source rather than the destination
+
+.. code-block:: ql
+
+    from FunctionCall strncpy, FunctionCall strlen, Expr sized
+    where
+      strncpy.getTarget().hasGlobalName("strncpy") and
+      strlen.getTarget().hasGlobalName("strlen") and
+      globalValueNumber(strncpy.getArgument(0)) != globalValueNumber(sized) and
+      globalValueNumber(strncpy.getArgument(1)) = globalValueNumber(sized) and
+      strlen = strncpy.getArgumen(2)
+      sized = strlen.getArgument(0)
+    select ci, "This call to strncpy is bounded by the size of the source rather than the destination"
+
+.. TODO: a second example
+
+Hash consing
+--------
+The hash consing library (defined in ``semmle.code.cpp.valuenumbering.HashCons``) provides a mechanism for identifying expressions that have the same syntactic structure. Hash consing is useful when your primary concern is with the text of the code. For instance, hash consing might be used to detect duplicate code within a function.
+
+The hash consing API
+~~~~~~~~~~~~~~~~~~~~
+The hash consing library exposes its interface primarily through the ``HashCons`` class. Each instance of ``HashCons`` represents a set of expressions that have the same syntax (including referring to the same variables). To get an expression in the set represented by a particular ``HashCons``, use the ``getAnExpr()`` member predicate.
+
+.. note::
+
+    While the ``HashCons`` class has ``toString`` and ``getLocation`` methods, these are only provided as debugging aids. They give the ``toString`` and ``getLocation`` of an arbitrary ``Instruction`` within the set.
+
+To get the ``HashCons`` of an ``Expr``, use the ``hashCons`` predicate.
+
+Examples
+~~~~~~~~
+
+.. TODO: prose explanations
+
+.. code-block:: ql
+
+    import cpp
+    import semmle.code.cpp.valuenumbering.HashCons
+
+    from IfStmt outer, IfStmt inner
+    where
+      outer.getElse+() = inner and
+      hashCons(outer.getCondition()) = hashCons(inner.getCondition())
+    select inner.getCondition(), "The condition of this if statement duplicates the condition of $@",
+      outer.getCondition(), "an enclosing if statement"
```

Comment:
Isn't GVN better for a query like this?

Reply:
It probably is - this was just the best `HashCons` query I could come up with. I'm open to suggestions.

Reply:
I still haven't thought of a query that's perfect for `HashCons`. The more I think about it, the more I think that `HashCons` is unexplainably floating in the middle between `GVN` at one end of the scale and comparison of raw parse trees at the other.